component/metrics/scrape/appendable.go

```diff
@@ -0,0 +1,89 @@
+package scrape
+
+import (
+	"context"
+	"sync"
+
+	"github.com/grafana/agent/component/metrics"
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// FlowMetric is a wrapper around a single sample without the timestamp.
+type FlowMetric struct {
+	Labels labels.Labels
+	Value  float64
+}
+
+type flowAppendable struct {
+	mut       sync.Mutex
+	buffer    map[int64][]*metrics.FlowMetric // Though mostly a map of 1 item, this allows it to work if more than one TS gets added
+	receivers []*metrics.Receiver
+}
+
+func newFlowAppendable(receivers ...*metrics.Receiver) *flowAppendable {
+	return &flowAppendable{
+		buffer:    make(map[int64][]*metrics.FlowMetric),
+		receivers: receivers,
+	}
+}
+
+func (app *flowAppendable) Appender(_ context.Context) storage.Appender {
+	return app
+}
```

Comment:
This will cause problems. 


The Prometheus scrape loop will request one appender per scrape job. If that scrape fails, it'll call .Rollback on the appender it received. 

If you used a shared appender for all scrape jobs, one failed scrape will cause the samples for all other metrics to be rolled back.

Reply:
Nice detail, I hadn't seen that! Hm, I'm trying to think of a solution here, by looking what the upstream code does.

- headAppender uses an atomically incremented appendID
- agent is using an AppenderPool 

I'll check if we're able to conditionally single out and rollback only the failed appender by using the shared timestamps, otherwise we can reuse one of the other solutions.

Reply:
Your suggestion was around this, right?https://github.com/grafana/agent/pull/1850/commits/f7ed6647aaf5319221bc3e3e2ed3281f3629112c

Reply:
Yep, looks good :)